### PR TITLE
Improve dockerfile directive parsing

### PR DESCRIPTION
**Pull request by tonistiigi**
Refactor so that new directives can be easily added in the future by just adding a new lines in https://github.com/tonistiigi/docker/blob/parse-directives/builder/dockerfile/parser/parser.go#L150-L152 . Simplify so that there is no need to store an instance of directives with the builder or pass it to the `Parse()` function. 

Fix bug where directive with whitespace prefix would still be applied.

In current version, error handling was inconsistent. Specifying invalid escape character would not produce an error but using escape directive twice would. Changed it so that if the directive has correct syntax and valid directive key, it will fail with a proper error.

cc @jhowardmsft

### CODE REVIEW

File: builder/dockerfile/parser/testfiles/escape_whitespace/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,5 @@
+  #escape = `
+
+FROM image
+ENV foo `
+ENV bar baz
```
Comment (duglin) [inline builder/dockerfile/parser/testfiles/escape_whitespace/Dockerfile]: github shows some funky chars at the end - is that due to a missing crlf?  Was that on purpose and part of the test? if so, let's add a comment so that people don't edit the file and add it back in by mistake.

File: builder/dockerfile/parser/testfiles/escape-nonewline/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# escape = ``
```
Comment (duglin) [inline builder/dockerfile/parser/testfiles/escape-nonewline/Dockerfile]: why was the char removed? Or, why did it work before?

Comment (tonistiigi) [inline builder/dockerfile/parser/testfiles/escape-nonewline/Dockerfile]: It's a bug in the previous implementation that allowed garbage data after the directive https://github.com/docker/docker/blob/61753568f5e482c551df6f3693cb7c942261b7de/builder/dockerfile/parser/parser.go#L51 . Only ` and \ are valid escape characters.

File: builder/dockerfile/parser/testfiles/escape_whitespace/Dockerfile (hunk 1 above)
Comment (tonistiigi) [inline builder/dockerfile/parser/testfiles/escape_whitespace/Dockerfile]: No, this doesn't change the test. Only thing we are testing is if the first line sets escape char or not.

File: builder/dockerfile/parser/parser.go [hunk 3]
```diff
@@ -32,31 +33,40 @@ type Node struct {
...
 // parsing directives.
-type Directive struct {
-	EscapeToken           rune           // Current escape token
-	LineContinuationRegex *regexp.Regexp // Current line contination regex
-	LookingForDirectives  bool           // Whether we are currently looking for directives
-	EscapeSeen            bool           // Whether the escape directive has been seen
+type Directives struct {
+	EscapeToken           rune                // Current escape token
+	LineContinuationRegex *regexp.Regexp      // Current line contination regex
+	usedDirectives        map[string]struct{} // Whether the escape directive has been seen
```
Comment (jhowardmsft) [inline builder/dockerfile/parser/parser.go]: I guess the comment could be updated here for this line.

File: builder/dockerfile/parser/testfiles/escape-nonewline/Dockerfile (hunk 2 above)
Comment (tonistiigi) [inline builder/dockerfile/parser/testfiles/escape-nonewline/Dockerfile]: @jhowardmsft It would be very easy to fix but I don't agree this is correct. Directive syntax should be `# key=value`, not `# key=value+any garbage`. Directives in general are not limited to single char.

File: builder/dockerfile/parser/testfiles/escape_whitespace/Dockerfile (hunk 1 above)
Comment (jhowardmsft) [inline builder/dockerfile/parser/testfiles/escape_whitespace/Dockerfile]: Wasn't deliberately missed.

File: builder/dockerfile/parser/parser.go [hunk 4]
```diff
@@ -148,66 +135,94 @@ func ParseLine(line string, d *Directive) (string, *Node, error) {
...
+			if err != nil {
+				d.usedDirectives[tecMatch[1]] = struct{}{}
+			}
+		}()
+		switch tecMatch[1] {
+		case "escape":
+			return d.SetEscapeToken(tecMatch[2])
+		}
+	}
+	return errNoDirective
```
Comment (jhowardmsft) [inline builder/dockerfile/parser/parser.go]: Is this returned externally? I don't believe we ever returned an error back to the caller on a bad directive, just silently swallowed it (by design).

File: builder/dockerfile/parser/parser.go [hunk 5]
```diff
@@ -32,31 +33,40 @@ type Node struct {
...
+		panic(err)
+	}
+	return *d
+}
+
 // SetEscapeToken sets the default token for escaping characters in a Dockerfile.
-func SetEscapeToken(s string, d *Directive) error {
+func (d *Directives) SetEscapeToken(s string) error {
 	if s != "`" && s != "\\" {
 		return fmt.Errorf("invalid ESCAPE '%s'. Must be ` or \\", s)
```
Comment (jhowardmsft) [inline builder/dockerfile/parser/parser.go]: I thought this was swallowed (by design) before and just ignored, the whole line being treated as a comment.

File: builder/dockerfile/parser/testfiles/escape-nonewline/Dockerfile (hunk 2 above)
Comment (jhowardmsft) [inline builder/dockerfile/parser/testfiles/escape-nonewline/Dockerfile]: I don't disagree with you in the slightest. Just commenting on why it was as it was - I know that bit came up in one of the many conversations when introducing the feature 😇

File: builder/dockerfile/parser/parser.go (hunk 5 above)
Comment (tonistiigi) [inline builder/dockerfile/parser/parser.go]: This causes a real error in current version and in this version.

File: builder/dockerfile/parser/parser.go [hunk 6]
```diff
@@ -32,31 +33,40 @@ type Node struct {
...
 )
 
 // DefaultEscapeToken is the default escape token
 const DefaultEscapeToken = "\\"
 
+// DefaultDirectives returns directives struct with default properties
+func DefaultDirectives() Directives {
+	d := &Directives{usedDirectives: make(map[string]struct{})}
+	if err := d.SetEscapeToken(DefaultEscapeToken); err != nil {
+		panic(err)
```
Comment (jhowardmsft) [inline builder/dockerfile/parser/parser.go]: Ok, but I'd still rather just log something.

File: builder/dockerfile/parser/parser.go (hunk 4 above)
Comment (tonistiigi) [inline builder/dockerfile/parser/parser.go]: This error is for catching the "silent error", if parsing reaches this error it will not be reported but parsing will just continue.

**Comment by thaJeztah** [comment]
Should we wait for https://github.com/docker/docker/pull/24725 to be merged? That one has been open for a long time

**Comment by jhowardmsft** [comment]
@thaJeztah No need - they're not at all related.

File: builder/dockerfile/parser/parser.go (hunk 3 above)
Comment (tonistiigi) [inline builder/dockerfile/parser/parser.go]: added

**Comment by jhowardmsft** [comment]
Seems reasonable to me. Moving to code review unless any objection from @duglin or @thaJeztah. Probably worth squashing the commits. My only real thing is the just-in-case `panic` but otherwise looks good.

File: integration-cli/docker_cli_build_test.go [hunk 7]
```diff
@@ -3577,8 +3577,8 @@ RUN [ "$(id -u):$(id -g)/$(id -un):$(id -gn)/$(id -G):$(id -Gn)" = '1001:1001/do
 
 # Switch back to root and double check that worked exactly as we might expect it to
 USER root
+# Add a "supplementary" group for our dockerio user
```
Comment (jhowardmsft) [inline integration-cli/docker_cli_build_test.go]: Should this be in a separate PR? Doesn't seem related...

**Comment by jhowardmsft** [comment]
Minor comments, LGTM when those are addressed.

File: integration-cli/docker_cli_build_test.go (hunk 7 above)
Comment (tonistiigi) [inline integration-cli/docker_cli_build_test.go]: It is needed. I thought I left a comment for that in this pr... . It is the same fix as in https://github.com/docker/docker/pull/24725/files#diff-64d1c3a97fd8b1abf170680883e81aaaR3580 . Doesn't matter how we handle empty lines, this Dockerfile should not be parseable.

**Comment by LK4D4** [comment]
@tonistiigi needs rebase
